docs/source/configuration.md
# Configuration

The _baseauth_ backend is configured through one `.env` file in the project root folder.
There is a template `env-skel` file available in the same folder, to copy from and
then modify it. See also [](install.md) on when and how to set up those file.

Environment variables without a default value are required to be set.

## `.env`

### Database Settings

These settings are used to set up the DB in the portfolio-postgres container.

#### `POSTGRES_DB`

Name of the PostgreSQL database.

#### `POSTGRES_USER`

User of the PostgreSQL database.

#### `POSTGRES_PASSWORD`

Password for user of the PostgreSQL database.

This needs to be set explicitly. Make sure to use some strong password, e.g.
one you can generate with `pwgen -sy 32 1`. You won't need to use it anywhere,
except if you want to connect directly to the database for debugging reasons.

#### `POSTGRES_PORT`

Default: `5432`

Port of the PostgreSQL database.

The database port only needs to be changed, if you are running Portfolio
locally in combination with e.g. Showroom also running locally. Then at
least one of the database container ports needs to be mapped to a different
value. So use whatever you set in your compose.override.yaml for
portfolio-postgres or use the default.

### Redis Settings

#### `REDIS_PORT`

Default: `6379`

Port of the Redis data store

Similar to `POSTGRES_PORT` you won't need to change this in most cases.
But if you develop locally and have several Redis instances for different
projects running at the same time, you might map some of them to alternate
ports. So whichever non-default port you set in your compose.override.yaml
for Redis, this should be also set here, unless you operate in a fully
containerised setup.

### DOCKER

For any online deployment the default (True) should be fine here, as usually all
services will be run inside docker containers. Only when you want to start the
Django application itself on your host machine (e.g. for local development), you have
to set this to False.

### SITE_URL & FORCE_SCRIPT_NAME

The `SITE_URL` has to be set to the base URL of the site _basauth_ is running on, which
will depend on whether you deploy this to some online node, either with multiple services
sharing one domain or running _baseauth_ on a separate domain, or if you run it locally.
For local development setups you can choose `http://127.0.0.1:8000/`. For an online
deployment choose the base path (protocol and domain), e.g. `https://base.uni-ak.ac.at/`.

Additionally, `FORCE_SCRIPT_NAME` (which defaults to `/auth`) will be used to
determine the actual PATH to your _baseauth_ instance, by prefixing it with the
`SITE_URL`. So for a local development setup (where Django is actually running on
127.0.0.1:8300) make sure to remove the comment and explicitly set this to an empty
string:

```
FORCE_SCRIPT_NAME=
```

Do the same if _baseauth_ runs on the root of a dedicated domain, and leave the
default if it runs on a shared domain where it runs on the _/auth_ path.

### BEHIND_PROXY

This defines whether your application is running behind a reverse proxy (e.g. nginx).
In most cases the default True will be fine here. But for local development you might
want to set this to False.

### DJANGO_ADMINS & DJANGO_SUPERUSERS

While `DJANGO_ADMINS` is only used to configure whom to send admin notifications,
`DJANGO_SUPERUSERS` tells Django which users to actually treat as administrators.
This is applied on every login, irrespective of which flags you set for single
users through the Django admin or by using the `createsuperuser` management command.
So if you create an admin user, make sure to add their username in this setting.
Otherwise, the user will automatically be demoted again.

### EMAIL\_\*

All settings in the block prefixed with `EMAIL_` are needed if you want to receive
e-Mail notifications from Django. While this is usually not necessary for local
development environments, it is highly advised for staging and production deployments.

### SESSION_COOKIE_DOMAIN

Usually you will use your baseauth server alongside some base applications. For the
relevant session cookies to work, we either need to run all applications on one
domain (in this case this setting can be left untouched), or we have to make sure
that the cookies provided by basauth to the client (the web browser) will be also
sent to the other base applications. For that you need to run the applications under
a common parent domain. In this case set this parameter to the common domain part.

E.g. if baseauth is running on baseauth.example.org and Portfolio is running on
portfolio.example.org, then set the SESSION_COOKIE_DOMAIN to example.org

For more background details:

- https://docs.djangoproject.com/en/4.2/ref/settings/#std-setting-SESSION_COOKIE_DOMAIN

### CORS\_\* & CSRF\_\*

For the frontend to work and being able to make authenticated requests on behalf
of the user you should minimally set `CORS_ALLOW_CREDENTIALS` to True. All other
settings should basically be fine by default, as long as your frontend runs on the
same domain as the backend. If you need frontends on different domains (e.g. for
testing and staging purposes) to be able to make those request, you should add them
to the `CSRF_TRUSTED_ORIGINS` and `CORS_ALLOWED_ORIGINS` lists.

Also in a setup where the different apps do not share the same domain you will have
to adapt the CSRF_COOKIE_DOMAIN, analog to the SESSION_COOKIE_DOMAIN.

For more background details on these settings:

- https://docs.djangoproject.com/en/4.2/ref/csrf/
- https://pypi.org/project/django-cors-headers/

### Showroom connection

These settings are only required, if a _Showroom_ instance is active, and syncing of
user profiles & preferences to _Showroom_ is activated. In order to activate syncing,
set `SYNC_TO_SHOWROOM` to `True`. Additionally, the `SHOWROOM_API_KEY` has to be
set to a key that has to be generated with the
[`create-source-repository`](https://showroom-backend.readthedocs.io/en/latest/management_commands.html#create-source-repository)
command, when _Showroom_ is set up. This key then can be used for _Portfolio_ and _baseauth_
to push new and updated activities and entities.

The other values only have to be set, if we deviate from a default single-domain
setup:

- `SHOWROOM_BASE_URL`: points to the base URL of where Showroom is available and
  defaults to `showroom/` appended to whatever is set as the `SITE_URL`.
- `SHOWROOM_API_PATH`: defaults to `api/v1/` and should only be changed if a new
  API version comes out

### Authentication backends

_baseauth_ can either be used as a standalone system, using Django's user model,
or it can use different authentication backends and/or a combination of those.

To configure which backends are used set the `AUTHENTICATION_BACKENDS` directive.
By default, this will only use Django's user model. In most system you might want
to replace this or add other backends in front of it.

Currently available options are:

- `django` - the internal user model authentication backend of Django
- `ldap` - a generic LDAP-based authentication
- `cas` - redirect all auth to an upstream CAS server

While `cas` will overrule other authentication backends, generally you can combine
different backends, so they will be checked in order of listing. So if
you for example use:

```
AUTHENTICATION_BACKENDS=ldap,django
```

_baseauth_ will first check against the configured LDAP directory, and only if it
cannot authenticate the user against it, will also check the Django user model. If
you switch those two parameters, the Django model will be checked first and LDAP
will only be checked if Django user does not exist. Be aware, that most authentication
backends create an internal Django user upon authentication. So in most cases it will
be most reasonable to put `django` last in the list - or even leave it out, if you do
not need local accounts (e.g. for testing oder administration).

Depending on which authentication backends you configure, you will also have to
add specific configuration directives. These will be explained in the following
subsections.

#### LDAP

If `ldap` is used in the `AUTHENTICATION_BACKENDS`, you also have to set the following:

- `AUTH_LDAP_SERVER_URI` : the server URI (including the port) of the LDAP server.
  E.g.: ldaps://ldap.example.org:636
- `AUTH_LDAP_BIND_DN` : the distinguished name used to query the directory
- `AUTH_LDAP_BIND_PASSWORD` : the password of the above user/DN
- `AUTH_LDAP_USER_ATTR_MAP` : a dictionary mapping the Django's internal user attributes
  to the LDAP user attributes. Defaults to `first_name=givenName,last_name=sn,email=mail`.

Now depending on your setup you can use three different ways to query your users:

1. Use `AUTH_LDAP_USER_DN_TEMPLATE` to set a string template that describes any
   user’s distinguished name based on the username. This must contain the placeholder
   %(user)s. E.g.: `uid=%(user)s,ou=users,dc=example,dc=com`. If this directive is
   set, the other directives are not relevant.
2. Use `AUTH_LDAP_USER_SEARCH_BASE` and `AUTH_LDAP_USER_SEARCH_USER_TEMPLATE` to
   configure the LDAP query. The first is the distinguished name of the search base.
   E.g.: `ou=users,dc=example,dc=com`. The second one is the filter string for the
   user lookup. E.g. `(cn=%(user)s)`
3. Similar to the second method, you can use the `AUTH_LDAP_USER_SEARCH_USER_TEMPLATE`
   as the filter string for the user lookup, but provide several distinguished names
   that should be used as the search base. These you can set as a semicolon-separated
   list of distinguished names. E.g.:
   `ou=unit1,ou=users,dc=example,dc=com;ou=unit2,ou=users,dc=example,dc=com`
   This will only work, if `AUTH_LDAP_USER_SEARCH_BASE` is not set.

#### CAS

```{note}
This authentication backend will overrule all other backends you might have configured.
All login and logout routes will be redirected to the upstream CAS server.
```

If `cas` is used in the `AUTHENTICATION_BACKENDS`, you also have to set the following:

- `CAS_SERVER` has to point to an upstream CAS server, which all login and logout
  attempts will be redirected to.

All other `CAS_` settings are optional (for details see the
[django-cas-ng config docs](https://djangocas.dev/docs/4.0/configuration.html)):

- `CAS_VERSION` : specifies the version of the CAS protocol and defaults to `3`
- `CAS_REDIRECT_URL` : specifies the redirect to happen after successful login and
  defaults to `/`. This only affects logins directly on baseauth. Logins coming
  from one of the apps already have their redirect URL set.
- `CAS_VERIFY_SSL_CERTIFICATE` : defaults to `True` and should not be changed in any
  production context. For local testing you can disable the SSL certificate check
  (e.g. if case of a self-signed certificate).
- `CAS_RENAME_ATTRIBUTES` : a dictionary used to rename the (key of the) attributes
  that the upstream CAS server may return. Is empty by default, so no renaming happens.
  For example, if `CAS_RENAME_ATTRIBUTES=ln=last_name,fn=first_name` the `ln` attribute
  returned by the cas server will be renamed as `last_name` and `fn` to `first_name`.


src/baseauth/settings.py
"""Django settings for CAS project.

Generated by 'django-admin startproject' using Django 2.2.3.

For more information on this file, see
https://docs.djangoproject.com/en/2.2/topics/settings/

For the full list of settings and their values, see
https://docs.djangoproject.com/en/2.2/ref/settings/

Before deployment please see
https://docs.djangoproject.com/en/2.2/howto/deployment/checklist/
"""

import os
import re
import sys
from email.utils import getaddresses
from pathlib import Path
from urllib.parse import urlparse

import environ
import ldap
from django_auth_ldap.config import LDAPSearch, LDAPSearchUnion

from django.urls import reverse_lazy
from django.utils.functional import lazy
from django.utils.translation import get_language, gettext_lazy as _

# Build paths inside the project like this: BASE_DIR / 'subdir'.
BASE_DIR = Path(__file__).resolve().parent.parent

PROJECT_NAME = '.'.join(__name__.split('.')[:-1])

env = environ.Env()
env.read_env(BASE_DIR / '..' / '.env')

try:
    from .secret_key import SECRET_KEY
except ImportError:
    from django.core.management.utils import get_random_secret_key

    with Path.open(BASE_DIR / PROJECT_NAME / 'secret_key.py', 'w+') as f:
        SECRET_KEY = get_random_secret_key()
        f.write(f"SECRET_KEY = '{SECRET_KEY}'\n")

DEFAULT_AUTO_FIELD = 'django.db.models.BigAutoField'

# SECURITY WARNING: don't run with debug turned on in production!
DEBUG = env.bool('DEBUG', default=False)

# Detect if executed under test
TESTING = any(
    test in sys.argv
    for test in (
        'test',
        'csslint',
        'jenkins',
        'jslint',
        'jtest',
        'lettuce',
        'pep8',
        'pyflakes',
        'pylint',
        'sloccount',
    )
)

DOCKER = env.bool('DOCKER', default=True)

SITE_URL = env.str('SITE_URL')

FORCE_SCRIPT_NAME = env.str('FORCE_SCRIPT_NAME', default='/auth')

SITE_HOSTNAME = urlparse(SITE_URL).hostname

ALLOWED_HOSTS = env.list('ALLOWED_HOSTS', default=[SITE_HOSTNAME])

BEHIND_PROXY = env.bool('BEHIND_PROXY', default=True)

DJANGO_ADMINS = env('DJANGO_ADMINS', default=None)

if DJANGO_ADMINS:
    ADMINS = getaddresses([DJANGO_ADMINS])
    MANAGERS = ADMINS

SUPERUSERS = env.tuple('DJANGO_SUPERUSERS', default=())


# Application definition

INSTALLED_APPS = [
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.staticfiles',
    # base common apps
    'base_common',
    'base_common_drf',
    # Third-party apps
    'mama_cas',
    'axes',
    'captcha',
    'corsheaders',
    'rest_framework',
    'rest_framework_api_key',
    'drf_spectacular',
    'easy_thumbnails',
    'django_rq',
    # Project apps
    'accounts',
    'core',
    'user_preferences',
    'showroom_connector',
]

AUTH_USER_MODEL = 'accounts.User'

# Authentication Backends
AUTH_BACKENDS_TO_USE = env.list('AUTHENTICATION_BACKENDS', default=['django'])

# Axes is always used to prevent brute-force attacks
AUTHENTICATION_BACKENDS = [
    'axes.backends.AxesBackend',
]

# Now all configurable authentication backends are added
for backend in AUTH_BACKENDS_TO_USE:
    # The default django user model backend
    if backend == 'django':
        AUTHENTICATION_BACKENDS.append('django.contrib.auth.backends.ModelBackend')

    if backend == 'cas':
        INSTALLED_APPS.append('django_cas_ng')
        AUTHENTICATION_BACKENDS.append('django_cas_ng.backends.CASBackend')

        CAS_SERVER_URL = env.str('CAS_SERVER', default=f'{SITE_URL}cas/')
        CAS_LOGIN_MSG = None
        CAS_LOGGED_MSG = None
        CAS_RENEW = False
        CAS_LOGOUT_COMPLETELY = True
        CAS_RETRY_LOGIN = True
        CAS_VERSION = env.str('CAS_VERSION', default='3')
        CAS_APPLY_ATTRIBUTES_TO_USER = True
        CAS_REDIRECT_URL = env.str('CAS_REDIRECT_URL', default=FORCE_SCRIPT_NAME or '/')
        CAS_VERIFY_SSL_CERTIFICATE = env.bool(
            'CAS_VERIFY_SSL_CERTIFICATE',
            default=True,
        )
        CAS_RENAME_ATTRIBUTES = env.dict('CAS_RENAME_ATTRIBUTES', default={})

    # A generically configurable LDAP authentication backend
    # See https://django-auth-ldap.readthedocs.io/en/latest/authentication.html for details
    if backend == 'ldap':
        AUTH_LDAP_CONNECTION_OPTIONS = {
            ldap.OPT_X_TLS_CACERTFILE: '/etc/ssl/certs/ca-certificates.crt',
            ldap.OPT_X_TLS_NEWCTX: 0,
        }

        AUTH_LDAP_SERVER_URI = env.str('AUTH_LDAP_SERVER_URI')
        AUTH_LDAP_BIND_DN = env.str('AUTH_LDAP_BIND_DN')
        AUTH_LDAP_BIND_PASSWORD = env.str('AUTH_LDAP_BIND_PASSWORD')
        try:
            AUTH_LDAP_USER_DN_TEMPLATE = env.str('AUTH_LDAP_USER_DN_TEMPLATE')
        except environ.ImproperlyConfigured:
            AUTH_LDAP_USER_SEARCH_USER_TEMPLATE = env.str(
                'AUTH_LDAP_USER_SEARCH_USER_TEMPLATE',
            )
            try:
                AUTH_LDAP_USER_SEARCH_BASE = env.str('AUTH_LDAP_USER_SEARCH_BASE')
                AUTH_LDAP_USER_SEARCH = LDAPSearch(
                    AUTH_LDAP_USER_SEARCH_BASE,
                    ldap.SCOPE_SUBTREE,
                    AUTH_LDAP_USER_SEARCH_USER_TEMPLATE,
                )
            except environ.ImproperlyConfigured:
                AUTH_LDAP_USER_SEARCH_BASE_LIST = env.str(
                    'AUTH_LDAP_USER_SEARCH_BASE_LIST',
                ).split(';')
                searches = [
                    LDAPSearch(
                        x.strip(),
                        ldap.SCOPE_SUBTREE,
                        AUTH_LDAP_USER_SEARCH_USER_TEMPLATE,
                    )
                    for x in AUTH_LDAP_USER_SEARCH_BASE_LIST
                ]
                AUTH_LDAP_USER_SEARCH = LDAPSearchUnion(*searches)

        AUTH_LDAP_USER_ATTR_MAP = env.dict(
            'AUTH_LDAP_USER_ATTR_MAP',
            default={'first_name': 'givenName', 'last_name': 'sn', 'email': 'mail'},
        )

        AUTH_LDAP_ALWAYS_UPDATE_USER = True
        AUTH_LDAP_CACHE_TIMEOUT = 0

        AUTHENTICATION_BACKENDS.append('django_auth_ldap.backend.LDAPBackend')

# CAS
MAMA_CAS_SERVICES = [
    {
        'SERVICE': rf'^http[s]?://{re.escape(urlparse(SITE_URL).hostname)}',
        'CALLBACKS': ['core.utils.get_attributes'],
        'LOGOUT_ALLOW': True,
        # 'LOGOUT_URL': '',
    },
]

MAMA_CAS_ENABLE_SINGLE_SIGN_OUT = True
"""Email settings."""
SERVER_EMAIL = f'error@{urlparse(SITE_URL).hostname}'

EMAIL_HOST_USER = env.str('EMAIL_HOST_USER', default='')
EMAIL_HOST_PASSWORD = env.str('EMAIL_HOST_PASSWORD', default='')
EMAIL_HOST = env.str('EMAIL_HOST', default='localhost')
EMAIL_PORT = env.int('EMAIL_PORT', default=25)
EMAIL_USE_TLS = env.bool('EMAIL_USE_TLS', default=False)
EMAIL_USE_LOCALTIME = env.bool('EMAIL_USE_LOCALTIME', default=True)

EMAIL_SUBJECT_PREFIX = '{} '.format(
    env.str('EMAIL_SUBJECT_PREFIX', default='[baseauth]').strip(),
)
if DEBUG:
    EMAIL_BACKEND = 'django.core.mail.backends.filebased.EmailBackend'
    EMAIL_FILE_PATH = BASE_DIR / '..' / 'tmp' / 'emails'

    if not EMAIL_FILE_PATH.exists():
        EMAIL_FILE_PATH.mkdir(parents=True)

""" Https settings """
if SITE_URL.startswith('https'):
    CSRF_COOKIE_SECURE = True
    SESSION_COOKIE_SECURE = True
    SECURE_SSL_REDIRECT = True
    SECURE_HSTS_SECONDS = 31536000

X_FRAME_OPTIONS = 'DENY'

MIDDLEWARE = [
    'base_common.middleware.HealthCheckMiddleware',
    'django.middleware.security.SecurityMiddleware',
    'whitenoise.middleware.WhiteNoiseMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.locale.LocaleMiddleware',
    'corsheaders.middleware.CorsMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.middleware.clickjacking.XFrameOptionsMiddleware',
    'axes.middleware.AxesMiddleware',
]
if 'cas' in AUTH_BACKENDS_TO_USE:
    MIDDLEWARE.append('django_cas_ng.middleware.CASMiddleware')

AXES_LOCKOUT_URL = reverse_lazy('locked_out')
AXES_VERBOSE = DEBUG

if BEHIND_PROXY:
    MIDDLEWARE += ['base_common.middleware.SetRemoteAddrFromForwardedFor']
    USE_X_FORWARDED_HOST = True
    SECURE_PROXY_SSL_HEADER = ('HTTP_X_FORWARDED_PROTO', 'https')

ROOT_URLCONF = f'{PROJECT_NAME}.urls'

TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': [BASE_DIR / 'templates'],
        'APP_DIRS': True,
        'OPTIONS': {
            'context_processors': [
                'django.template.context_processors.debug',
                'django.template.context_processors.request',
                'django.contrib.auth.context_processors.auth',
                'django.contrib.messages.context_processors.messages',
            ],
            'debug': DEBUG,
            'string_if_invalid': "[invalid variable '%s'!]" if DEBUG else '',
        },
    },
]

WSGI_APPLICATION = f'{PROJECT_NAME}.wsgi.application'


# Database
# https://docs.djangoproject.com/en/4.2/ref/settings/#databases

POSTGRES_HOST_DEFAULT = f'{PROJECT_NAME}-postgres' if DOCKER else 'localhost'
DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.postgresql',
        'NAME': env.str('POSTGRES_DB', default=f'django_{PROJECT_NAME}'),
        'USER': env.str('POSTGRES_USER', default=f'django_{PROJECT_NAME}'),
        'PASSWORD': env.str('POSTGRES_PASSWORD'),
        'HOST': env.str('POSTGRES_HOST', default=POSTGRES_HOST_DEFAULT),
        'PORT': env.str('POSTGRES_PORT', default='5432'),
    },
}


# Password validation
# https://docs.djangoproject.com/en/2.2/ref/settings/#auth-password-validators

AUTH_PASSWORD_VALIDATORS = [
    {
        'NAME': (
            'django.contrib.auth.password_validation.UserAttributeSimilarityValidator'
        ),
    },
    {'NAME': 'django.contrib.auth.password_validation.MinimumLengthValidator'},
    {'NAME': 'django.contrib.auth.password_validation.CommonPasswordValidator'},
    {'NAME': 'django.contrib.auth.password_validation.NumericPasswordValidator'},
]


# Internationalization
# https://docs.djangoproject.com/en/2.2/topics/i18n/

LANGUAGE_CODE = 'en'
TIME_ZONE = 'Europe/Vienna'
USE_I18N = True
USE_L10N = True
USE_TZ = True

LANGUAGES = (('de', _('German')), ('en', _('English')))

LANGUAGES_DICT = dict(LANGUAGES)

LOCALE_PATHS = [
    BASE_DIR / 'locale',
    BASE_DIR / 'locale_mama_cas',
]

get_language_lazy = lazy(get_language, str)


# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/2.2/howto/static-files/

STATICFILES_DIRS = ('{}{}'.format(os.path.normpath(BASE_DIR / 'static'), os.sep),)
STATIC_URL = '{}/s/'.format(FORCE_SCRIPT_NAME if FORCE_SCRIPT_NAME else '')
STATIC_ROOT = '{}{}'.format(
    os.path.normpath(BASE_DIR / 'assets' / 'static'),
    os.sep,
)

MEDIA_URL = '{}/m/'.format(FORCE_SCRIPT_NAME if FORCE_SCRIPT_NAME else '')
MEDIA_ROOT = '{}{}'.format(
    os.path.normpath(BASE_DIR / 'assets' / 'media'),
    os.sep,
)

FILE_UPLOAD_PERMISSIONS = 0o644

if FORCE_SCRIPT_NAME:
    WHITENOISE_STATIC_PREFIX = '/s/'
STATICFILES_STORAGE = 'whitenoise.storage.CompressedManifestStaticFilesStorage'


# Logging
LOG_DIR = BASE_DIR / '..' / 'logs'

if not LOG_DIR.exists():
    LOG_DIR.mkdir(parents=True)

LOGGING = {
    'version': 1,
    'disable_existing_loggers': False,
    'formatters': {
        'verbose': {
            'format': (
                '%(levelname)s %(asctime)s %(module)s %(process)d %(thread)d '
                '%(message)s'
            ),
        },
        'simple': {'format': '%(levelname)s %(message)s'},
        'simple_with_time': {'format': '%(levelname)s %(asctime)s %(message)s'},
    },
    'handlers': {
        'null': {'level': 'DEBUG', 'class': 'logging.NullHandler'},
        'console': {
            'level': 'DEBUG',
            'class': 'logging.StreamHandler',
            'formatter': 'simple',
        },
        'file': {
            'level': 'DEBUG',
            'class': 'concurrent_log_handler.ConcurrentTimedRotatingFileHandler',
            'filename': LOG_DIR / 'application.log',
            'when': 'midnight',
            'backupCount': 1000,
            'use_gzip': True,
            'formatter': 'verbose',
        },
        'mail_admins': {
            'level': 'ERROR',
            'class': 'django.utils.log.AdminEmailHandler',
        },
        'stream_to_console': {'level': 'DEBUG', 'class': 'logging.StreamHandler'},
        'rq_console': {
            'level': 'DEBUG',
            'class': 'logging.StreamHandler',
            'formatter': 'simple_with_time',
        },
    },
    'loggers': {
        '': {
            'handlers': ['console', 'file', 'mail_admins'],
            'propagate': True,
            'level': 'INFO',
        },
        'django': {
            'handlers': ['console', 'file', 'mail_admins'],
            'propagate': True,
            'level': 'INFO',
        },
        'django.request': {
            'handlers': ['console', 'file', 'mail_admins'],
            'level': 'ERROR',
            'propagate': True,
        },
        'rq.worker': {
            'handlers': ['rq_console', 'mail_admins'],
            'level': 'DEBUG',
            'propagate': False,
        },
    },
}


# Cache settings
CACHES = {
    'default': {
        'BACKEND': 'django_redis.cache.RedisCache',
        'LOCATION': 'redis://{}:{}/0'.format(
            f'{PROJECT_NAME}-redis' if DOCKER else 'localhost',
            env.str('REDIS_PORT', default='6379'),
        ),
        'OPTIONS': {'CLIENT_CLASS': 'django_redis.client.DefaultClient'},
    },
}


# RQ worker settings
RQ_QUEUES = {
    'default': {'USE_REDIS_CACHE': 'default', 'DEFAULT_TIMEOUT': 500},
}

if DEBUG or TESTING:
    for queue_config in iter(RQ_QUEUES.values()):
        queue_config['ASYNC'] = False

RQ_EXCEPTION_HANDLERS = ['core.rq.handlers.exception_handler']

RQ_FAILURE_TTL = 2628288  # approx. 3 month


# Session settings
SESSION_ENGINE = 'django.contrib.sessions.backends.cache'
SESSION_CACHE_ALIAS = 'default'
SESSION_COOKIE_NAME = f'sessionid_{PROJECT_NAME}'
SESSION_COOKIE_DOMAIN = env.str('SESSION_COOKIE_DOMAIN', default=None)
SESSION_EXPIRE_AT_BROWSER_CLOSE = True

CSRF_COOKIE_NAME = f'csrftoken_{PROJECT_NAME}'
CSRF_COOKIE_DOMAIN = env.str('CSRF_COOKIE_DOMAIN', default=None)
CSRF_TRUSTED_ORIGINS = env.list('CSRF_TRUSTED_ORIGINS', default=[])

CORS_ALLOW_CREDENTIALS = env.bool('CORS_ALLOW_CREDENTIALS', default=False)
CORS_ALLOW_ALL_ORIGINS = env.bool('CORS_ALLOW_ALL_ORIGINS', default=False)
CORS_ALLOWED_ORIGINS = env.list('CORS_ALLOWED_ORIGINS', default=[])

HOME_REDIRECT = reverse_lazy('cas_login')


# Axes settings
AXES_FAILURE_LIMIT = 3
AXES_COOLOFF_TIME = 1  # number in hours

CAPTCHA_FLITE_PATH = '/usr/bin/flite'

if DEBUG:
    INTERNAL_IPS = ('127.0.0.1',)

    if 'ldap' in AUTH_BACKENDS_TO_USE:
        LOGGING['loggers']['django_auth_ldap'] = {
            'level': 'DEBUG',
            'handlers': ['console'],
        }


# User profile picture thumbnail settings
THUMBNAIL_NAMER = 'easy_thumbnails.namers.hashed'
THUMBNAIL_DIR = 'tn'
THUMBNAIL_MEDIA_ROOT = f'{os.path.normpath(Path(MEDIA_ROOT) / THUMBNAIL_DIR)}{os.sep}'
THUMBNAIL_MEDIA_URL = f'{MEDIA_URL}/{THUMBNAIL_DIR}/'
THUMBNAIL_QUALITY = 95
THUMBNAIL_OPTIONS = {
    'size': (625, 625),
    'crop': 'smart',
}

MAX_IMAGE_SIZE = 10485760  # 10mb


# User preferences settings profiles for apps
SETTINGS_DATA = {
    'showroom': {
        'title': 'Showroom',
        'icon': 'https://base.uni-ak.ac.at/bs/img/icons/showroom.svg',
        'settings': {
            'activate_profile': {
                'title': {
                    'de': 'Aktivieren Sie Ihre persönliche Showroom Seite',
                    'en': 'Activate your personal Showroom page',
                },
                'type': 'boolean',
                'default_value': False,
            },
        },
    },
}


# API specific settings
SWAGGER_SETTINGS = {
    'SECURITY_DEFINITIONS': {
        'X-Api-Key': {'type': 'apiKey', 'in': 'header', 'name': 'X-Api-Key'},
    },
}

API_KEY_CUSTOM_HEADER = 'HTTP_X_API_KEY'

REST_FRAMEWORK = {
    'DEFAULT_SCHEMA_CLASS': 'drf_spectacular.openapi.AutoSchema',
    'DEFAULT_AUTHENTICATION_CLASSES': (
        'rest_framework.authentication.SessionAuthentication',
    ),
    'DEFAULT_PERMISSION_CLASSES': ['rest_framework.permissions.IsAuthenticated'],
    'DEFAULT_RENDERER_CLASSES': ('rest_framework.renderers.JSONRenderer',),
    'DEFAULT_VERSIONING_CLASS': 'rest_framework.versioning.URLPathVersioning',
    'DEFAULT_VERSION': 'v1',
    'ORDERING_PARAM': 'sort',
    'DEFAULT_FILTER_BACKENDS': ['django_filters.rest_framework.DjangoFilterBackend'],
}

SPECTACULAR_SETTINGS = {
    'TITLE': 'baseauth',
    'DESCRIPTION': """Provides authenticated access only to a user's preferences and
    profile data, as well as API-Key based authenticated access to user data for user
    data agents (e.g. in Showroom). The repo and further documentation are at
    [https://github.com/base-angewandte/baseauth](https://github.com/base-angewandte/baseauth).
    """,
    'VERSION': '1.0',
    'SERVE_INCLUDE_SCHEMA': False,
    'COMPONENT_SPLIT_REQUEST': True,
    'TAGS': ['autosuggest', 'user', 'users'],
    'SERVERS': [
        {
            'url': env.str(
                'OPENAPI_SERVER_URL',
                default=f'{SITE_URL.rstrip("/")}{FORCE_SCRIPT_NAME}',
            ),
            'description': env.str(
                'OPENAPI_SERVER_DESCRIPTION',
                default='baseauth / User Preferences',
            ),
        },
    ],
    # available SwaggerUI configuration parameters
    # https://swagger.io/docs/open-source-tools/swagger-ui/usage/configuration/
    'SWAGGER_UI_SETTINGS': {
        'displayOperationId': True,
    },
}


# Autosuggest
SKOSMOS_API = 'https://voc.uni-ak.ac.at/skosmos/rest/v1/'
TAX_ID = 'potax'
TAX_GRAPH = 'http://base.uni-ak.ac.at/portfolio/taxonomy/'
VOC_ID = 'povoc'
VOC_GRAPH = 'http://base.uni-ak.ac.at/portfolio/vocabulary/'
LANGUAGES_VOCID = 'languages'

ACCEPT_LANGUAGE_HEADER = {'Accept-Language': get_language_lazy()}

GND_MAPPING = {
    'source': 'id',  # common_schema: GND schema
    'label': 'label',
}

VIAF_CONTRIBUTORS_MAPPING = {
    'label': 'displayForm',
}

CONTRIBUTORS = (
    'GND_PERSON',
    'VIAF_PERSON',
)

REQUESTS_TIMEOUT = 60


# Showroom connection settings
SYNC_TO_SHOWROOM = env.bool('SYNC_TO_SHOWROOM', default=False)
SHOWROOM_BASE_URL = env.str('SHOWROOM_BASE_URL', default=f'{SITE_URL}showroom/')
SHOWROOM_API_PATH = env.str('SHOWROOM_API_PATH', default='api/v1/')
SHOWROOM_API_BASE = f'{SHOWROOM_BASE_URL}{SHOWROOM_API_PATH}'
SHOWROOM_API_KEY = env.str('SHOWROOM_API_KEY', default=None)
WORKER_DELAY = 3


# Sentry
SENTRY_DSN = env.str('SENTRY_DSN', default=None)
SENTRY_ENVIRONMENT = env.str(
    'SENTRY_ENVIRONMENT',
    default='development'
    if any(i in SITE_URL for i in ['dev', 'localhost', '127.0.0.1'])
    else 'production',
)
SENTRY_TRACES_SAMPLE_RATE = env.float('SENTRY_TRACES_SAMPLE_RATE', default=0.2)

if SENTRY_DSN:
    import sentry_sdk
    from sentry_sdk.integrations.django import DjangoIntegration
    from sentry_sdk.integrations.redis import RedisIntegration
    from sentry_sdk.integrations.rq import RqIntegration

    sentry_sdk.init(
        dsn=SENTRY_DSN,
        environment=SENTRY_ENVIRONMENT,
        integrations=[
            DjangoIntegration(),
            RedisIntegration(),
            RqIntegration(),
        ],
        traces_sample_rate=SENTRY_TRACES_SAMPLE_RATE,
        send_default_pii=True,
    )
